src/forwardproblem.cpp

```diff
@@ -41,29 +42,118 @@ ForwardProblem::ForwardProblem(
     : model(model)
     , solver(solver)
     , edata(edata)
-    , nroots_(gsl::narrow<decltype(nroots_)::size_type>(model->ne), 0)
-    , rootvals_(gsl::narrow<decltype(rootvals_)::size_type>(model->ne), 0.0)
-    , rval_tmp_(gsl::narrow<decltype(rval_tmp_)::size_type>(model->ne), 0.0)
     , dJzdx_(model->nJ * model->nx_solver * model->nMaxEvent(), 0.0)
     , t_(model->t0())
-    , roots_found_(model->ne, 0)
-    , x_(model->nx_solver, solver->getSunContext())
-    , x_old_(model->nx_solver, solver->getSunContext())
-    , dx_(model->nx_solver, solver->getSunContext())
-    , xdot_(model->nx_solver, solver->getSunContext())
-    , xdot_old_(model->nx_solver, solver->getSunContext())
-    , sx_(model->nx_solver, model->nplist(), solver->getSunContext())
-    , sdx_(model->nx_solver, model->nplist(), solver->getSunContext())
-    , stau_(model->nplist())
-    , uses_presimulation_(edata && edata->t_presim > 0) {}
+    , uses_presimulation_(edata && edata->t_presim > 0)
+    , ws_(model, solver)
+    , main_simulator_(model, solver, &ws_, &dJzdx_)
+    , pre_simulator_(model, solver, &ws_, &dJzdx_) {}
+
+void EventHandlingSimulator::run(
+    realtype const t0, ExpData const* edata,
+    std::vector<realtype> const& timepoints
+) {
+    std::ranges::fill(ws_->nroots, 0);
+
+    t0_ = t0;
+    t_ = t0;
+    ws_->tlastroot = t0;
+
+    // handle initial events
+    if (model_->ne && std::ranges::any_of(ws_->roots_found, [](int rf) {
+            return rf == 1;
+        })) {
+        handleEvent(true, edata);
+    }
+
+    // store initial state and sensitivity
+    result.initial_state_ = getSimulationState();
```

Comment:
Would argue that we should store initial state post assignment but pre event execution. Different situation with PEtab converted timecourses, but we may want to have special handling/reporting there anyways.

Reply:
I also stumbled across that, but this is how it was always done, it's only more obvious now. I mainly didn't change it because I wasn't sure whether it breaks something downstream.

I think one can argue for either. On the one hand, if the initial state is defined as $x(t_0)$, it should be post-event. It may be somewhat confusing if $x_0 = x(t_0)$ and "initial state" differ. On the other hand, we have `Model::fx0`, which in the case of events may not yield the actual `x_0`, and get/setInitialState which is currently pre-event.
It feels more intuitive to consider everything that happens at $t = t_0$ as initialization, but this is not really compatible with the rest...

No strong opinion. I can change it and try to clarify that in the documentation.